src/AI/JakeMan.java

```diff
@@ -846,29 +1107,122 @@ else if( um == copter ) // Consider downgrading a tank to an inf to get a copter
+        allCounterContestants.add(ratio.counter);
+    }
+
     if( null == copter ) // I clearly don't understand this unit set, so just grab something to hedge
       copter = myArmy.cos[0].getUnitModel(UnitModel.AIR_TO_AIR, false);
 
     capPhase = new CapPhaseAnalyzer(map, myArmy);
   }
 
+  private void counterBuildSetup()
+  {
+    counterOrder = new ArrayList<>();
+    unitTypeToCounterMap = new HashMap<>();
+    counterTypeNegationMap = new HashMap<>(); // Note: we assume all UnitModels referenced in here are in the above
+    if( !myInfo.buildCounters )
+      return;
+
+    // tanks: no need to calc them beyond the Md clause
+    // copters: 1.5 copters or 1AA per
+    if( null != copter )
+    {
+      counterOrder.add(copter);
+      unitTypeToCounterMap.put(copter, new ArrayList<>());
+      var copterCounters = unitTypeToCounterMap.get(copter);
+      CounterRatio copterCounterCopter = new CounterRatio(copter, 200/3);
+      copterCounterCopter.roundTargetPercentUp = true; // I don't consider a single existing copter a full counter to an enemy copter
+      copterCounters.add(copterCounterCopter);
+      copterCounters.add(new CounterRatio(antiAir, 100));
+    }
+
+    // Md: 1 neo per 1.5 Mds? Ignore if you already have a bomber
+    counterOrder.add(mdTank);
+    unitTypeToCounterMap.put(mdTank, new ArrayList<>());
+    var mdCounters = unitTypeToCounterMap.get(mdTank);
+    var bomber = myArmy.cos[0].getUnitModel(UnitModel.AIR_TO_SURFACE | UnitModel.JET, false);
+    if( null != bomber )
+      mdCounters.add(new CounterRatio(bomber, 250));
+    var neoTank = allTanks.get(2);
+    if( null != neoTank )
+      mdCounters.add(new CounterRatio(neoTank, 150));
+
+    // 1 fighter per 1.7 (yes) bombers
+    //   (trunctate after the multiplicarion, so you get one in response to one bomber and a second in response to the third),
+    var fighter = myArmy.cos[0].getUnitModel(UnitModel.AIR_TO_AIR | UnitModel.JET, false);
+    if( null != bomber )
+    {
+      counterOrder.add(bomber);
+      unitTypeToCounterMap.put(bomber, new ArrayList<>());
+      var bomberCounters = unitTypeToCounterMap.get(bomber);
+      if( null != fighter )
+        bomberCounters.add(new CounterRatio(fighter, 170));
+      bomberCounters.add(new CounterRatio(antiAir, 50));
+
+      // or 2 non-copter-calc-involved AAs per bomber
+      if( null != copter )
+      {
+        counterTypeNegationMap.put(bomber, new HashMap<>());
+        var bomberCounterNegators = counterTypeNegationMap.get(bomber);
+        bomberCounterNegators.put(antiAir, new ArrayList<>());
+        var aaCounterCounters = bomberCounterNegators.get(antiAir);
+        aaCounterCounters.add(new CounterRatio(copter, 100));
+      }
+    }
+
+    // stealth: have 1 healthy fighter on the board if a stealth is present
+    var stealths = myArmy.cos[0].getAllModels(UnitModel.AIR_TO_SURFACE | UnitModel.AIR_TO_AIR | UnitModel.JET, false);
```

Comment:
This is still a super awkward way to identify a specific unit type.

Reply:
Would you prefer a search by name? lol
The intent is to default to searching by *relevant properties*, so you aren't stuck with e.g. not affecting Neotanks or whatever.

Reply:
Sure, I get wanting to be able to get a unit or list of units based on shared properties. On the other hand, it's also super convenient to be able to just grab a specific unit type, which is what you are trying to do here.
Besides which, the existence of a way to grab a specific unit type in no way prevents using attribute-based filtering. The real argument for this being the "preferred" mechanism is that it allows us to make the code apply across all unit sets, but then you still have code like `mdTank = allTanks.get(1);`, which betrays that pretense.

Reply:
> On the other hand, it's also super convenient to be able to just grab a specific unit type, which is what you are trying to do here.

I suppose. I guess it'd be possible to make a universal unit enum that doesn't interact with the damage tables (since that would be a lot of bloat).
But it feels weird to set up something like that with what's ultimately pretty niche utility? And you'd still have to check for absence of that unit type, which IMO the current API makes a little more obvious.
> then you still have code like `mdTank = allTanks.get(1);`, which betrays that pretense.

I dunno. It does a semi-sane thing in KaijuWars instead of exploding.
It's obviously not _ideal_, but this AI isn't designed with even DoR in mind, let alone that.

Reply:
> I guess it'd be possible to make a universal unit enum

One with values like TANK_SMALL, TANK_MD, TANK_BIG or whatever, that may or may not exist for a given unit set? I think I might lean toward (for cases like Jake where you have UnitSet-specific logic) to have a function that just explicitly sets his internal tank/mdTank/etc variables to the right UnitModel based on which UnitSet is active (and maybe fall back to the current logic if the UnitSet isn't the expected one?). It would consolidate the mess, though it's debatable whether it's a real improvement.

> It does a semi-sane thing in KaijuWars instead of exploding.

Sure, but it's far from a general solution, though a general solution is admittedly a hard ask here.

Reply:
> I think I might lean toward (for cases like Jake where you have UnitSet-specific logic) to have a function that just explicitly sets his internal tank/mdTank/etc variables to the right UnitModel based on which UnitSet is active (and maybe fall back to the current logic if the UnitSet isn't the expected one?).

So what, it would downcast the unit types and check against the damage chart enum values? Or maybe "downcast" like the visiting that happens for weapons?
I assume you're not proposing the the unit set expose a function specifically for picking units for JakeMan.
> Sure, but it's far from a general solution, though a general solution is admittedly a hard ask here.

`ASSAULT` in flag terms is basically "essence of tank", and we've made the assumption that unit sets should order their units in rough cost/practicality order.
I think it's about as general as it's getting (without digging into the damage table algorithmically).